cachito/workers/pkg_managers/npm.py

```diff
@@ -27,6 +30,455 @@
+        self.is_top_level = is_top_level
+        self.alias = alias
+        self.path = path
+        self.dependent_packages = dependent_packages or []
+
+    @property
+    def version(self) -> str:
+        """Get the package version.
+
+        For v1/v2 package-lock.json `dependencies`, this will be a semver
+        for registry dependencies and a url for git/https/filepath sources.
+        https://docs.npmjs.com/cli/v6/configuring-npm/package-lock-json#dependencies
+        For v2+ package-lock.json `packages`, this will be a semver from the package.json file.
+        https://docs.npmjs.com/cli/v7/configuring-npm/package-lock-json#packages
+        """
+        return self._package_dict["version"]
+
+    @version.setter
+    def version(self, version: str) -> None:
+        """Set the package version."""
+        self._package_dict["version"] = version
+
+    @property
+    def resolved_url(self) -> str:
+        """Get the location where the package was resolved from.
+
+        For v1/v2 package-lock.json `dependencies`, this will be the "resolved"
+        key for registry deps and the "version" key for non-registry deps.
+        For v2+ package-lock.json `packages`, this will be the "resolved" key
+        unless it is a file dep, in which case it will be the path to the file.
+        """
+        if self.path is not None and "resolved" not in self._package_dict:
+            return f"file:{self.path}"
+
+        return self._package_dict.get("resolved") or self.version
+
+    def set_resolved(self, resolved: str) -> None:
+        """Set the location where the package was resolved from."""
+        self._package_dict["resolved"] = resolved
+        # The "from" value is the original value from package.json for some
+        # locations. Remove it while setting a new `resolved` location
+        self._package_dict.pop("from", None)
+
+    @property
+    def bundled(self) -> bool:
+        """Return True if this package is bundled."""
+        return any(self._package_dict.get(key) for key in ["bundled", "inBundle"])
+
+    @property
+    def dev(self) -> bool:
+        """Return True if this package is a dev dependency."""
+        return any(self._package_dict.get(key) for key in ("dev", "devOptional"))
+
+    @property
+    def is_link(self) -> bool:
+        """Return True if this package is a link."""
+        return self._package_dict.get("link", False)
+
+    @property
+    def integrity(self) -> Optional[str]:
+        """Get the package subresource integrity string."""
+        return self._package_dict.get("integrity")
+
+    @integrity.setter
+    def integrity(self, integrity: str) -> None:
+        """Set the package subresource integrity string."""
+        self._package_dict["integrity"] = integrity
+
+    @property
+    def is_file_dep(self) -> bool:
+        """Return True if this package is a file dependency."""
+        return self.resolved_url.startswith("file:") and not self.bundled
```

Comment:
Hmm, so bundled deps get a `file:` resolved url? That's a little misleading, but then again there's no un-misleading resolved url for bundled deps :thinking: 

Except maybe `None`, but I expect that would result in a larger refactoring. I can live with the current behavior